// @effect/platform/dist/cjs/OpenApiJsonSchema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeWithDefs = exports.make = exports.fromAST = void 0;
var JSONSchema = _interopRequireWildcard(require("effect/JSONSchema"));
var Record = _interopRequireWildcard(require("effect/Record"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @since 1.0.0
 */

/**
 * @category encoding
 * @since 1.0.0
 */
const make = schema => {
  const defs = {};
  const out = makeWithDefs(schema, {
    defs
  });
  if (!Record.isEmptyRecord(defs)) {
    out.$defs = defs;
  }
  return out;
};
/**
 * Creates a schema with additional options and definitions.
 *
 * **Options**
 *
 * - `defs`: A record of definitions that are included in the schema.
 * - `defsPath`: The path to the definitions within the schema (defaults to "#/$defs/").
 * - `topLevelReferenceStrategy`: Controls the handling of the top-level reference. Possible values are:
 *   - `"keep"`: Keep the top-level reference (default behavior).
 *   - `"skip"`: Skip the top-level reference.
 * - `additionalPropertiesStrategy`: Controls the handling of additional properties. Possible values are:
 *   - `"strict"`: Disallow additional properties (default behavior).
 *   - `"allow"`: Allow additional properties.
 *
 * @category encoding
 * @since 1.0.0
 */
exports.make = make;
const makeWithDefs = (schema, options) => fromAST(schema.ast, options);
/** @internal */
exports.makeWithDefs = makeWithDefs;
const fromAST = (ast, options) => {
  const jsonSchema = JSONSchema.fromAST(ast, {
    definitions: options.defs,
    definitionPath: options.defsPath ?? "#/components/schemas/",
    target: "openApi3.1",
    topLevelReferenceStrategy: options.topLevelReferenceStrategy,
    additionalPropertiesStrategy: options.additionalPropertiesStrategy
  });
  return jsonSchema;
};
exports.fromAST = fromAST;
